File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int k = Integer.parseInt(st.nextToken());
		int n = Integer.parseInt(st.nextToken());

		long[] lengths = new long[k];
		long max = 0;

		for (int i = 0; i < k; i++) {
			lengths[i] = Long.parseLong(br.readLine());
			max = Math.max(max, lengths[i]);
		}

		long left = 1;
		long right = max;
		long answer = 0;

		while (left <= right) {
			long mid = (left + right) / 2;
			long count = 0;

			for (int i = 0; i < k; i++) {
				count += lengths[i] / mid;
			}

			if (count >= n) {
				answer = Math.max(answer, mid);
				left = mid + 1;
			} else {
				right = mid - 1;
			}
		}

		System.out.println(answer);
	}
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

k, n = map(int, sys.stdin.readline().rstrip().split())
length_lst = [int(sys.stdin.readline().rstrip()) for _ in range(k)]

left = 1
right = max(length_lst)
answer = 0

while left <= right:
    mid = (left + right) // 2
    cnt = 0

    for i in range(k):
        cnt += length_lst[i] // mid

    if cnt >= n:
        answer = max(answer, mid)
        left = mid + 1
    else:
        right = mid - 1

print(answer)
